common/configurable/src/main/java/io/helidon/common/configurable/ThreadPool.java

```diff
@@ -468,20 +471,24 @@ private static WorkQueue createQueue(int capacity, int corePoolSize, int maxPool
     /**
      * A queue that tracks peak and average sizes.
      */
-    static class WorkQueue extends LinkedBlockingQueue<Runnable> {
+    static class WorkQueue extends ConcurrentLinkedQueue<Runnable> implements BlockingQueue<Runnable> {
         private final int capacity;
         private final LongAdder totalSize;
         private final AtomicInteger totalTasks;
         private final AtomicInteger peakSize;
+        private final Semaphore semaphoreRead;
+        private final Semaphore semaphoreWrite;
 
         /**
-         * Constructor.
+         * Constructor. Initially {@code capacity} writes (enqueues) and 0 reads (dequeues)
+         * are available.
          *
          * @param capacity The queue capacity.
          */
         WorkQueue(int capacity) {
-            super(capacity);
             this.capacity = capacity;
+            this.semaphoreRead = new Semaphore(0);
+            this.semaphoreWrite = new Semaphore(capacity);
```

Comment:
Is there a reason (I'm sure there is) that a `ReadWriteLock` isn't used instead?

Reply:
That's just a pair of locks, right? I don't think that would enable the type of concurrency that is sought after here unless I'm missing something.

Reply:
Two things:

1. `Semaphore` allows up to `capacity` concurrent writers to make progress. (A lock showed contention)
2. `Semaphore` allows a release by a thread that doesn't own the lock. (readers unblock writers, writers unblock readers)

Reply:
The second point was what I wasn't remembering; thanks!